pkg/apis/configuration/validation/validation.go

```diff
@@ -707,6 +738,130 @@ func validateVariable(nVar string, validVars map[string]bool, fieldPath *field.P
 		msg := fmt.Sprintf("'%v' contains an invalid NGINX variable. Accepted variables are: %v", nVar, mapToPrettyString(validVars))
 		allErrs = append(allErrs, field.Invalid(fieldPath, nVar, msg))
 	}
+	return allErrs
+}
+
+func isValidSpecialVariableHeader(header string) []string {
+	errMsgs := validation.IsHTTPHeaderName(strings.Replace(header, "_", "-", -1))
+	if len(errMsgs) >= 1 || strings.Contains(header, "-") {
+		return []string{"a valid HTTP header must consist of alphanumeric characters or '_'"}
+	}
+	return nil
+}
+
+func validateSpecialVariable(nVar string, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	value := strings.SplitN(nVar, "_", 2)
+
+	if strings.HasPrefix(nVar, "arg") {
```

Comment:
Let me know what you think of the following suggestion:

validate every special variable in its own function.
* isValidSpecialVariableHeader
* isValidSpecialVariableArgument
* isValidSpecialVariableCookie

Each of those function can have its own regexes - I think it would be better instead of modying the input and calling other functions, even if we have some duplication.

Also, each of those function will return a standardize special variable error message. For example: '$http_asdf-sdf must contain ... '

I think that will simplify the code and make it more readable.

Also, if a validation variable returns a slice of string, for consistency, I suggest using the `for _, msg := range` approach instead of `if len(validationMsg) >= 1 `

Reply:
This sounds reasonable, I was trying to reduce number of functions as this file is growing and growing, but I guess is fine.

Reply:
PS: about regexes, I'll give it a thought, I am not a fan of regexes all the time, it's easier to make a mistake with a regular expression and do not cover all the cases than with string manipulation in my opinion.